adding StackFrameExtensions apis

mairaw opened the pull request:
@joperezr, @wtgodbe would you be the right reviewers for this class?

[Internal review URL](https://review.docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframeextensions?branch=pr-en-us-3261&view=netcore-2.0)

mairaw committed:
adding StackFrameExtensions apis

joperezr commented:
This one in particular is not very well defined in terms of ownership that I'm aware of, but I can definitely help taking a look.

joperezr commented on xml/System.Diagnostics/StackFrameExtensions.xml:
```diff
@@ -46,10 +46,11 @@
-        <returns>To be added.</returns>
+        <param name="stackFrame">A stack frame.</param>
+        <summary>Returns a pointer to the base address of the native just-in-time (JIT)-compiled image that this stack frame is executing.</summary>
+        <returns>This method always returns <see cref="F:System.IntPtr.Zero" />.</returns>
```
I believe this is not entirely true for .NET Native case. @trylek is that right?

joperezr commented on xml/System.Diagnostics/StackFrameExtensions.xml:
```diff
@@ -73,9 +74,9 @@
-        <returns>To be added.</returns>
+        <param name="stackFrame">A stack frame.</param>
+        <summary>Gets an interface pointer to the start of the native code for the method that is being executed.</summary>
+        <returns>This method always returns <see cref="F:System.IntPtr.Zero" />.</returns>
```
Same Here, if you take a look [here](https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Diagnostics/StackFrameExtensions.cs) you'll see that in .Net Native we actually return something else.

joperezr commented:
Left a couple of comments but looks good in general.

BillWagner reviewed (approved):
Overall, this looks great. Once you address @joperezr comments, :shipit:

trylek commented on xml/System.Diagnostics/StackFrameExtensions.xml:
```diff
@@ -154,10 +159,12 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="stackFrame">A stack frame.</param>
+        <summary>Indicates whether the native just-in-time (JIT)-compiled image is available for the specified stack frame.</summary>
```
I think that in similar cases it would be easiest to just remove the direct reference to JIT assuming the sentence continues making sense - e.g. here "Indicates whether the native image..."

trylek commented:
@mairaw Apologies for not responding earlier. When I was originally adding the APIs, I believe I copied the XML comments from the .NETStandard 2 docs; admittedly these were heavily geared towards CoreCLR execution model, thus all the mentions of JITted code.

It is true that on AOT runtimes (.NET Native, CoreRT) their semantics is slightly different - in particular, the "start of native module" corresponds to the load address of the AOT-compiled binary module containing the executing method and native IP offsets within methods correspond to the precompiled, not JITted, native code. Other than that, semantic differences should be minimal, at least I hope so.

As far as the documentation goes, it seems to me that it would be easiest to fix the bits mentioned by @jopererz (I have also spotted one) by minimum editation involving removal of the reference to JIT; we should typically end up with just a mention of a "native image" or "native module" as that is the common denominator of both JITted and AOT runtimes.

Thanks

Tomas

rpetrusha commented:
@mairaw, when you address the comments from @trylek, this will be ready to merge.

mairaw committed:
Merge branch 'master' into StackFrameExtensions

mairaw committed:
feedback

mairaw commented:
Thank you for your feedback @joperezr and @trylek. I believe I've addressed all your comments. Can you please take a look? (The PR looks a bit funky now because the content that was missing from MSDN was ported in bulk since I've created this).

Once it finishes building, you can see the built content here:
https://review.docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframeextensions?branch=pr-en-us-3261

Apologies for the delay in addressing the feedback!

mairaw commented:
@joperezr @trylek does this look good now? Thanks!

trylek reviewed (approved):
LGTM, thanks Maira!
